File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/context/ExecutionContext.java
package ru.kmorozov.gbd.core.logic.context;

import ru.kmorozov.db.core.config.IContextLoader;
import ru.kmorozov.gbd.core.logic.Proxy.AbstractProxyListProvider;
import ru.kmorozov.gbd.core.logic.Proxy.HttpHostExt;
import ru.kmorozov.gbd.core.logic.Proxy.UrlType;
import ru.kmorozov.gbd.core.logic.extractors.base.AbstractHttpProcessor;
import ru.kmorozov.gbd.core.logic.extractors.base.IImageExtractor;
import ru.kmorozov.gbd.core.logic.extractors.base.IPostProcessor;
import ru.kmorozov.gbd.logger.Logger;
import ru.kmorozov.gbd.logger.consumers.AbstractOutputReceiver;
import ru.kmorozov.gbd.logger.progress.IProgress;
import ru.kmorozov.gbd.utils.QueuedThreadPoolExecutor;

import java.util.*;
import java.util.concurrent.TimeUnit;

/**
 * Created by km on 22.11.2015.
 */
public final class ExecutionContext {

    static final Logger logger = Logger.getLogger(ExecutionContext.class);

    private static final String EMPTY = "";
    public static ExecutionContext INSTANCE;
    private final boolean singleMode;
    private final AbstractOutputReceiver output;
    private final Map<String, BookContext> bookContextMap = new HashMap<>();
    public QueuedThreadPoolExecutor<BookContext> bookExecutor;
    public QueuedThreadPoolExecutor<BookContext> pdfExecutor;

    private ExecutionContext(AbstractOutputReceiver output, boolean singleMode) {
        this.output = output;
        this.singleMode = singleMode;
    }

    public static synchronized void initContext(AbstractOutputReceiver output, boolean singleMode) {
        ExecutionContext.INSTANCE = new ExecutionContext(output, singleMode);
    }

    public Logger getLogger(Class<?> clazz, BookContext bookContext) {
        return Logger.getLogger(this.output, clazz.getName(), this.singleMode || null == bookContext ? ExecutionContext.EMPTY : bookContext.getBookInfo().getBookData().getTitle() + ": ");
    }

    public Logger getLogger(Class<?> clazz) {
        return this.getLogger(clazz, null);
    }

    public Logger getLogger(String location) {
        return Logger.getLogger(this.output, location, ExecutionContext.EMPTY);
    }

    public void addBookContext(IBookListProducer idsProducer, IProgress progress, IPostProcessor postProcessor) {
        for (String bookId : idsProducer.getBookIds()) {
            try {
                this.bookContextMap.computeIfAbsent(bookId, k -> new BookContext(bookId, progress, postProcessor));
            } catch (Exception ex) {
                ExecutionContext.logger.severe("Cannot add book " + bookId + " because of " + ex.getMessage());
            }
        }
    }

    public List<BookContext> getContexts(boolean shuffle) {
        List<BookContext> contexts = Arrays.asList(this.bookContextMap.values().toArray(new BookContext[0]));
        if (shuffle) Collections.shuffle(contexts);
        return contexts;
    }

    public AbstractOutputReceiver getOutput() {
        return this.output;
    }

    public static int getProxyCount() {
        return AbstractProxyListProvider.getInstance().getProxyCount();
    }

    public synchronized void updateProxyList() {
        AbstractProxyListProvider.getInstance().updateProxyList();
    }

    public synchronized void updateBlacklist() {
        AbstractProxyListProvider.updateBlacklist();
    }

    public void execute() {
        this.bookExecutor = new QueuedThreadPoolExecutor<>((long) this.bookContextMap.size(), 5, BookContext::isImgStarted, "bookExecutor");
        this.pdfExecutor = new QueuedThreadPoolExecutor<>((long) this.bookContextMap.size(), 5, BookContext::isPdfCompleted, "pdfExecutor");

        for (BookContext bookContext : this.getContexts(true)) {
            IImageExtractor extractor = bookContext.getExtractor();
            extractor.newProxyEvent(HttpHostExt.NO_PROXY);
            this.bookExecutor.execute(extractor);
        }

        AbstractProxyListProvider.getInstance().processProxyList(UrlType.GOOGLE_BOOKS);

        this.bookExecutor.terminate(10L, TimeUnit.MINUTES);
        this.pdfExecutor.terminate(30L, TimeUnit.MINUTES);

        long totalProcessed = this.getContexts(false).stream().mapToLong(BookContext::getPagesProcessed).sum();
        this.getLogger("Total").info("Total pages processed: " + totalProcessed);

        final IContextLoader contextProvider = ContextProvider.getContextProvider();

        contextProvider.updateIndex();
        contextProvider.updateContext();
        this.updateBlacklist();
        AbstractHttpProcessor.close();
    }

    public static void sendProxyEvent(HttpHostExt proxy) {
        if (ExecutionContext.INSTANCE != null)
            ExecutionContext.INSTANCE.newProxyEvent(proxy);
    }

    private void newProxyEvent(HttpHostExt proxy) {
        for (BookContext bookContext : this.getContexts(true))
            bookContext.getExtractor().newProxyEvent(proxy);
    }

    public void postProcessBook(BookContext bookContext) {
        this.pdfExecutor.execute(bookContext.getPostProcessor());
    }

    public boolean isSingleMode() {
        return this.singleMode;
    }

    public Iterable<String> getBookIds() {
        return this.bookContextMap.keySet();
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/context/ExecutionContext.kt
package ru.kmorozov.gbd.core.logic.context

import ru.kmorozov.gbd.core.config.GBDOptions
import ru.kmorozov.gbd.core.logic.extractors.base.AbstractHttpProcessor
import ru.kmorozov.gbd.core.logic.extractors.base.IPostProcessor
import ru.kmorozov.gbd.core.logic.library.ILibraryMetadata
import ru.kmorozov.gbd.core.logic.library.LibraryFactory
import ru.kmorozov.gbd.core.logic.library.metadata.GoogleBooksMetadata.Companion.GOOGLE_METADATA
import ru.kmorozov.gbd.core.logic.library.metadata.UnknownMetadata
import ru.kmorozov.gbd.core.logic.proxy.HttpHostExt
import ru.kmorozov.gbd.core.logic.proxy.providers.AbstractProxyListProvider
import ru.kmorozov.gbd.logger.Logger
import ru.kmorozov.gbd.logger.output.ReceiverProvider
import ru.kmorozov.gbd.utils.QueuedThreadPoolExecutor
import java.net.InetSocketAddress
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

/**
 * Created by km on 22.11.2015.
 */
class ExecutionContext private constructor(val isSingleMode: Boolean) {
    var defaultMetadata: ILibraryMetadata = GOOGLE_METADATA

    val bookIds: Iterable<String>
        get() = bookContextMap.keys

    fun getLogger(clazz: Class<*>, bookContext: BookContext? = null): Logger {
        return Logger.getLogger(ReceiverProvider.getReceiver(GBDOptions.debugEnabled), clazz.name, if (isSingleMode || null == bookContext) EMPTY else bookContext.bookInfo.bookData.title + ": ")
    }

    fun addBookContext(idsProducer: IBookListProducer, postProcessor: IPostProcessor) {
        idsProducer.bookIds.stream().parallel().forEach {
            try {
                bookContextMap.computeIfAbsent(it) { BookContext(it, postProcessor) }
            } catch (ex: Exception) {
                logger.severe("Cannot add book $it because of $ex.message")
            }
        }
    }

    fun getContexts(shuffle: Boolean): List<BookContext> {
        val contexts = mutableListOf(*bookContextMap.values.toTypedArray())
        if (shuffle) contexts.shuffle()
        return contexts
    }

    @Synchronized
    fun updateProxyList() {
        AbstractProxyListProvider.INSTANCE.updateProxyList()
    }

    @Synchronized
    fun updateBlacklist() {
        AbstractProxyListProvider.updateBlacklist()
    }

    fun size(): Int {
        return bookContextMap.size
    }

    fun execute() {
        val contexts = getContexts(true)

        bookExecutor.reset(contexts.size)
        pdfExecutor.reset(contexts.size)

        for (bookContext in contexts) {
            if (bookContext.isPdfCompleted) {
                bookContext.started.set(false)
                bookContext.pdfCompleted.set(false)
            }

            bookContext.resetBookInfo()
            val extractor = bookContext.extractor
            extractor.reset()

            if (GBDOptions.serverMode)
                AbstractProxyListProvider.INSTANCE.parallelProxyStream.forEach { extractor.newProxyEvent(it) }

            bookExecutor.execute(extractor)
        }

        defaultMetadata = LibraryFactory.getMetadata(contexts)
        if (defaultMetadata == UnknownMetadata.UNKNOWN_METADATA) {
            for (bookContext in contexts)
                postProcessBook(bookContext)

            pdfExecutor.reset(0)
            bookExecutor.terminate(10L, TimeUnit.MINUTES)
            pdfExecutor.terminate(30L, TimeUnit.MINUTES)

            AbstractHttpProcessor.close()
        } else {
            bookExecutor.terminate(10L, TimeUnit.MINUTES)
            pdfExecutor.terminate(30L, TimeUnit.MINUTES)

            val totalProcessed = getContexts(false).stream().mapToLong({ x -> (x as BookContext).pagesProcessed.get() }).sum()
            getLogger("Total").info("Total pages processed: $totalProcessed")

            val contextProvider = ContextProvider.contextProvider

            contextProvider.updateContext()
            updateBlacklist()
            AbstractHttpProcessor.close()
        }
    }

    fun forceComepleteOne() {
        pdfExecutor.dec()
    }

    fun forceCompleteAll() {
        bookContextMap.values.forEach { it.forceComplete() }
    }

    private fun newProxyEvent(proxy: HttpHostExt) {
        for (bookContext in getContexts(true))
            bookContext.extractor.newProxyEvent(proxy)
    }

    fun postProcessBook(bookContext: BookContext) {
        if (!bookContext.isEmpty)
            pdfExecutor.execute(bookContext.getPostProcessor())
    }

    fun inProcess(): Boolean {
        return bookContextMap.values.filter { !it.pdfCompleted.get() }.count() > 0
    }

    companion object {
        private val bookContextMap: MutableMap<String, BookContext> = ConcurrentHashMap<String, BookContext>()
        lateinit var bookExecutor: QueuedThreadPoolExecutor<BookContext>
        lateinit var pdfExecutor: QueuedThreadPoolExecutor<BookContext>
        lateinit var proxyExecutor: QueuedThreadPoolExecutor<InetSocketAddress>

        internal val logger = Logger.getLogger(GBDOptions.debugEnabled, ExecutionContext::class.java)

        private const val EMPTY = ""
        lateinit var INSTANCE: ExecutionContext

        var initialized = false

        @Synchronized
        fun initContext(singleMode: Boolean) {
            INSTANCE = ExecutionContext(singleMode)

            bookExecutor = QueuedThreadPoolExecutor(bookContextMap.size, 5, { it.isImgStarted || it.isEmpty }, "bookExecutor")
            pdfExecutor = QueuedThreadPoolExecutor(bookContextMap.size, 5, { it.isPdfCompleted || it.isEmpty }, "pdfExecutor")

            initialized = true
        }

        fun getLogger(clazz: Class<*>): Logger {
            return Logger.getLogger(ReceiverProvider.getReceiver(GBDOptions.debugEnabled), clazz.name, "")
        }

        fun getLogger(location: String): Logger {
            return Logger.getLogger(ReceiverProvider.getReceiver(), location, EMPTY)
        }

        val proxyCount: Int
            get() = AbstractProxyListProvider.INSTANCE.proxyCount

        fun sendProxyEvent(proxy: HttpHostExt) {
            INSTANCE.newProxyEvent(proxy)
        }
    }
}